subsys/net/lib/tls_credentials/tls_credentials_trusted.c

```diff
@@ -456,3 +460,399 @@ int tls_credential_delete(sec_tag_t tag, enum tls_credential_type type)
 
 	return ret;
 }
+
+#if defined(CONFIG_TLS_CREDENTIAL_KEYGEN)
+
+#include <psa/crypto.h>
+#include <mbedtls/pk.h>
+#include <mbedtls/asn1.h>
+
+/* For this backend, the private-key will be stored in DER format.
+ * Use `cred get <tag> PK bin` to retrieve (the non-terminated base64 encoding of) the key using
+ * the credential shell.
+ */
+int tls_credential_keygen(sec_tag_t tag, enum tls_credential_keygen_type type,
+			  void *key_buf, size_t *key_len)
+{
+	int ret = 0;
+
+	/* PSA keygen */
+	psa_status_t psa_status;
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	psa_key_id_t key_id = PSA_KEY_ID_NULL;
+	psa_key_usage_t key_usage;
+
+	/* MbedTLS key formatting */
+	int mbed_status;
+	mbedtls_pk_context pk_ctx;
+	void* key_start = 0;
+	size_t key_max;
+
+	/* Permanent key storage */
+	int cred_status;
+	psa_storage_uid_t uid = tls_credential_get_uid(tag, TLS_CREDENTIAL_PRIVATE_KEY);
+	bool key_stored = false;
+
+	if (key_len == NULL) {
+		return -EINVAL;
+	}
+
+	key_max = *key_len;
+
+	k_mutex_lock(&credential_lock, K_FOREVER);
+	mbedtls_pk_init(&pk_ctx);
+
+	/* Ensure the storage destination is valid. */
+	if (!tag_type_valid(tag, TLS_CREDENTIAL_PRIVATE_KEY)) {
+		ret = -EINVAL;
+		goto cleanup;
+	}
+
+	/* Presently, only SECP256R1 is supported by this backend for keygen */
+	if (!tls_credential_can_keygen(type)) {
+		ret = -ENOTSUP;
+		goto cleanup;
+	}
+
+	/* Verify sectag not already taken */
+	if (tls_credential_toc_find_slot(uid) != CRED_MAX_SLOTS) {
+		ret = -EEXIST;
+		goto cleanup;
+	}
+
+	/* Before attempting keygen, check that an empty slot is available */
+	if (tls_credential_toc_find_slot(0) == CRED_MAX_SLOTS) {
+		ret = -ENOMEM;
+		goto cleanup;
+	}
+
+	/* Use PSA to generate a volatile SECP256R1 private/public key-pair
+         *
+	 * It is possible (and better) to mark the private key as persistent, and store it directly
```

Comment:
Since @tomi-font 's PR about secure storage has been merged (#76222) you can now store the private key in the Internal Trusted Storage.

Reply:
Fantastic -- We still don't have support for it, though, in the Zephyr TLS stack. That will be in future commits. For now it still needs to be stored via the old mechanism